File: pcc/py_runtime/src/py_process_timeout.c
/* Transitional subprocess timeout helper for the C and pcc-Python runtimes.
 *
 * The child gets its own process group so timeout cleanup reaches
 * grandchildren as well as the immediate command. Platform wait/signal
 * ownership is already routed to freestanding pcc-Python in the production
 * archive; spawn and argv construction remain to migrate.
 */

#include "py_internal.h"

#include <errno.h>
#include <fcntl.h>
#include <signal.h>
#include <spawn.h>
#include <stdint.h>
#include <stdlib.h>
#include <string.h>
#include <sys/wait.h>
#include <time.h>
#include <unistd.h>

extern char **environ;

#define PCC_SUBPROCESS_TIMEOUT_RC (-124)
#define PCC_TIMEOUT_POLL_NS 10000000L
#define PCC_TIMEOUT_TERM_GRACE_MS 200

static void free_exec_argv(char **items, int64_t count) {
    if (items == NULL) return;
    for (int64_t i = 0; i < count; i++) free(items[i]);
    free(items);
}

static char **build_exec_argv(PyObject *argv, int64_t *count_out) {
    int64_t count = py_obj_len(argv);
    if (count <= 0) return NULL;
    char **items = (char **)calloc((size_t)count + 1, sizeof(char *));
    if (items == NULL) return NULL;

    for (int64_t i = 0; i < count; i++) {
        PyObject *index = py_int_from_i64(i);
        PyObject *item = py_obj_getitem(argv, index);
        py_decref(index);
        PyObject *text = py_obj_str(item);
        py_decref(item);
        if (text == NULL) {
            free_exec_argv(items, i);
            return NULL;
        }
        const char *raw = py_str_utf8(text);
        size_t size = raw != NULL ? strlen(raw) : 0;
        items[i] = (char *)malloc(size + 1);
        if (items[i] == NULL) {
            py_decref(text);
            free_exec_argv(items, i);
            return NULL;
        }
        if (size > 0) memcpy(items[i], raw, size);
        items[i][size] = '\0';
        py_decref(text);
    }
    items[count] = NULL;
    *count_out = count;
    return items;
}

static int64_t monotonic_millis(void) {
    int64_t now_us = pcc_runtime_monotonic_us();
    return now_us > 0 ? now_us / 1000 : -1;
}

#ifndef PCC_USE_FREESTANDING_PLATFORM_PROCESS
int64_t py_process_normalize_wait_status(int64_t raw_status) {
    if (raw_status == -1) return 127;
    int status = (int)raw_status;
    if (WIFEXITED(status)) return (int64_t)WEXITSTATUS(status);
    if (WIFSIGNALED(status)) return -(int64_t)WTERMSIG(status);
    return 127;
}
#endif

static int64_t runtime_waitpid(pid_t pid, int *status, int options) {
#ifdef PCC_USE_FREESTANDING_PLATFORM_PROCESS
    return pcc_platform_waitpid((int64_t)pid, (int32_t *)status,
                                (int64_t)options);
#else
    pid_t waited;
    do {
        waited = waitpid(pid, status, options);
    } while (waited < 0 && errno == EINTR);
    return (int64_t)waited;
#endif
}

static int64_t runtime_kill(pid_t pid, int signal_number) {
#ifdef PCC_USE_FREESTANDING_PLATFORM_PROCESS
    return pcc_platform_kill((int64_t)pid, (int64_t)signal_number);
#else
    return (int64_t)kill(pid, signal_number);
#endif
}

static int wait_for_exit(pid_t pid, int *status, int64_t deadline_ms) {
    for (;;) {
        int64_t waited = runtime_waitpid(pid, status, WNOHANG);
        if (waited == (int64_t)pid) return 1;
        if (waited < 0) return -1;
        int64_t now_ms = monotonic_millis();
        if (now_ms < 0 || now_ms >= deadline_ms) return 0;
        (void)pcc_runtime_sleep_ns(PCC_TIMEOUT_POLL_NS);
    }
}

static void terminate_process_group(pid_t pid, int *status) {
    /* POSIX_SPAWN_SETPGROUP made pid the process-group id. Keep the direct-pid
     * signal as a defensive fallback if a platform rejects group delivery. */
    if (runtime_kill(-pid, SIGTERM) != 0) runtime_kill(pid, SIGTERM);
    int64_t now_ms = monotonic_millis();
    int64_t deadline_ms = now_ms < 0 ? 0 : now_ms + PCC_TIMEOUT_TERM_GRACE_MS;
    int waited = wait_for_exit(pid, status, deadline_ms);
    if (waited == 1 || waited < 0) return;

    if (runtime_kill(-pid, SIGKILL) != 0) runtime_kill(pid, SIGKILL);
    (void)runtime_waitpid(pid, status, 0);
}

int64_t py_subprocess_run_timeout(
    PyObject *argv,
    int32_t capture_output,
    int64_t timeout_ms
) {
    if (timeout_ms <= 0) return 127;

    int64_t count = 0;
    char **items = build_exec_argv(argv, &count);
    if (items == NULL) return 127;

    posix_spawn_file_actions_t actions;
    posix_spawnattr_t attr;
    int actions_ready = posix_spawn_file_actions_init(&actions) == 0;
    int attr_ready = posix_spawnattr_init(&attr) == 0;
    if (!actions_ready || !attr_ready) {
        if (actions_ready) posix_spawn_file_actions_destroy(&actions);
        if (attr_ready) posix_spawnattr_destroy(&attr);
        free_exec_argv(items, count);
        return 127;
    }

    int setup_error = 0;
    if (capture_output != 0) {
        setup_error = posix_spawn_file_actions_addopen(
            &actions, STDOUT_FILENO, "/dev/null", O_WRONLY, 0
        );
        if (setup_error == 0) {
            setup_error = posix_spawn_file_actions_addopen(
                &actions, STDERR_FILENO, "/dev/null", O_WRONLY, 0
            );
        }
    }
    if (setup_error == 0) {
        setup_error = posix_spawnattr_setpgroup(&attr, 0);
    }
    if (setup_error == 0) {
        setup_error = posix_spawnattr_setflags(&attr, POSIX_SPAWN_SETPGROUP);
    }

    pid_t pid = -1;
    int spawn_error = setup_error;
    if (spawn_error == 0) {
#ifdef PCC_USE_FREESTANDING_PLATFORM_ENV
        char **spawn_env = pcc_platform_env_snapshot();
        if (spawn_env == NULL) {
            spawn_error = ENOMEM;
        } else {
            spawn_error = posix_spawnp(
                &pid, items[0], &actions, &attr, items, spawn_env
            );
            pcc_platform_env_snapshot_free(spawn_env);
        }
#else
        spawn_error = posix_spawnp(
            &pid, items[0], &actions, &attr, items, environ
        );
#endif
    }
    posix_spawn_file_actions_destroy(&actions);
    posix_spawnattr_destroy(&attr);
    free_exec_argv(items, count);
    if (spawn_error != 0) return 127;

    int64_t start_ms = monotonic_millis();
    if (start_ms < 0) {
        int status = 0;
        terminate_process_group(pid, &status);
        return 127;
    }
    int status = 0;
    int waited = wait_for_exit(pid, &status, start_ms + timeout_ms);
    if (waited == 1) return py_process_normalize_wait_status(status);
    if (waited < 0) return 127;

    terminate_process_group(pid, &status);
    return PCC_SUBPROCESS_TIMEOUT_RC;
}

File: pcc/py_runtime/py/py_process_timeout.py
"""pcc-Python owner of subprocess timeout and process-group cleanup."""

__pcc_runtime_port__ = True

from pcc.extern import c_abi_export, c_int64, c_ptr, c_void, extern
from pcc.unsafe import (
    free,
    load_i32,
    load_ptr,
    malloc,
    memcpy,
    null,
    ptr_is_null,
    store_i32,
    store_i8,
    store_ptr,
    strlen,
    unsigned_div_i64,
)


py_decref = extern("py_decref", (c_ptr,), c_void)
py_int_from_i64 = extern("py_int_from_i64", (c_int64,), c_ptr)
py_obj_getitem = extern("py_obj_getitem", (c_ptr, c_ptr), c_ptr)
py_obj_len = extern("py_obj_len", (c_ptr,), c_int64)
py_obj_str = extern("py_obj_str", (c_ptr,), c_ptr)
py_str_utf8 = extern("py_str_utf8", (c_ptr,), c_ptr)

platform_env_snapshot = extern("pcc_platform_env_snapshot", (), c_ptr)
platform_env_snapshot_free = extern(
    "pcc_platform_env_snapshot_free", (c_ptr,), c_void
)
platform_spawnp = extern(
    "pcc_platform_spawnp", (c_ptr, c_ptr, c_int64), c_int64
)
platform_waitpid = extern(
    "pcc_platform_waitpid", (c_int64, c_ptr, c_int64), c_int64
)
platform_kill = extern("pcc_platform_kill", (c_int64, c_int64), c_int64)
platform_monotonic_us = extern("pcc_platform_monotonic_us", (), c_int64)
platform_sleep_ns = extern("pcc_platform_sleep_ns", (c_int64,), c_int64)
normalize_wait_status = extern(
    "py_process_normalize_wait_status", (c_int64,), c_int64
)


def _free_exec_argv(items, count: int) -> None:
    if ptr_is_null(items):
        return
    index = 0
    while index < count:
        free(load_ptr(items, index * 8))
        index = index + 1
    free(items)


def _build_exec_argv(argv):
    count = py_obj_len(argv)
    if count <= 0 or count > 1048576:
        return null()
    items = malloc((count + 1) * 8)
    if ptr_is_null(items):
        return null()
    index = 0
    while index < count:
        py_index = py_int_from_i64(index)
        item = py_obj_getitem(argv, py_index)
        py_decref(py_index)
        text = py_obj_str(item)
        py_decref(item)
        if ptr_is_null(text):
            _free_exec_argv(items, index)
            return null()
        raw = py_str_utf8(text)
        size = 0
        if not ptr_is_null(raw):
            size = strlen(raw)
        owned = malloc(size + 1)
        if ptr_is_null(owned):
            py_decref(text)
            _free_exec_argv(items, index)
            return null()
        if size > 0:
            memcpy(owned, raw, size)
        store_i8(owned, size, 0)
        store_ptr(items, index * 8, owned)
        py_decref(text)
        index = index + 1
    store_ptr(items, count * 8, null())
    return items


def _monotonic_millis() -> int:
    now_us = platform_monotonic_us()
    if now_us <= 0:
        return -1
    return unsigned_div_i64(now_us, 1000)


def _wait_for_exit(pid: int, status, deadline_ms: int) -> int:
    while True:
        waited = platform_waitpid(pid, status, 1)
        if waited == pid:
            return 1
        if waited < 0:
            return -1
        now_ms = _monotonic_millis()
        if now_ms < 0 or now_ms >= deadline_ms:
            return 0
        # Runtime-library modules are linked without running their Python
        # module initializer, so ABI constants must remain literal here rather
        # than being loaded from uninitialized module-global storage.
        platform_sleep_ns(10000000)


def _terminate_process_group(pid: int, status) -> None:
    if platform_kill(-pid, 15) != 0:
        platform_kill(pid, 15)
    now_ms = _monotonic_millis()
    deadline_ms = 0
    if now_ms >= 0:
        deadline_ms = now_ms + 200
    waited = _wait_for_exit(pid, status, deadline_ms)
    if waited == 1 or waited < 0:
        return
    if platform_kill(-pid, 9) != 0:
        platform_kill(pid, 9)
    platform_waitpid(pid, status, 0)


@c_abi_export("py_subprocess_run_timeout")
def py_subprocess_run_timeout(
    argv, capture_output: int, timeout_ms: int
) -> int:
    if timeout_ms <= 0:
        return 127
    count = py_obj_len(argv)
    items = _build_exec_argv(argv)
    if ptr_is_null(items):
        return 127
    child_env = platform_env_snapshot()
    if ptr_is_null(child_env):
        _free_exec_argv(items, count)
        return 127
    pid = platform_spawnp(items, child_env, capture_output)
    platform_env_snapshot_free(child_env)
    _free_exec_argv(items, count)
    if pid <= 0:
        return 127

    # waitpid writes one C ``int``.  Keep the raw slot at the platform ABI
    # width instead of over-allocating it as though it were an int64 result.
    status = malloc(4)
    if ptr_is_null(status):
        return 127
    store_i32(status, 0, 0)
    start_ms = _monotonic_millis()
    if start_ms < 0:
        _terminate_process_group(pid, status)
        free(status)
        return 127
    waited = _wait_for_exit(pid, status, start_ms + timeout_ms)
    if waited == 1:
        result = normalize_wait_status(load_i32(status, 0))
        free(status)
        return result
    if waited < 0:
        free(status)
        return 127
    _terminate_process_group(pid, status)
    free(status)
    return -124
